scripts/bert_SQuAD/dataset.py

```diff
@@ -0,0 +1,493 @@
+                    if self.is_training:
+
+                        if self.version_2:
+                            is_impossible = qa['is_impossible']
+                        if (len(qa['answers']) != 1) and (not is_impossible):
+                            raise ValueError(
+                                'For training, each question should have exactly 1 answer.'
+                            )
+                        if not is_impossible:
+                            answer = qa['answers'][0]
+                            orig_answer_text = answer['text']
+                            answer_offset = answer['answer_start']
+                            answer_length = len(orig_answer_text)
+                            start_position = char_to_word_offset[answer_offset]
+                            end_position = char_to_word_offset[
+                                answer_offset + answer_length - 1]
+                            # Only add answers where the text can be exactly recovered from the
+                            # document. If this CAN'T happen it's likely due to weird Unicode
+                            # stuff so we will just skip the example.
+                            #
+                            # Note that this means for training mode, every example is NOT
+                            # guaranteed to be preserved.
+                            actual_text = ' '.join(
+                                doc_tokens[start_position:(end_position + 1)])
+                            cleaned_answer_text = ' '.join(
+                                self.whitespace_tokenize(orig_answer_text))
+                            if actual_text.find(cleaned_answer_text) == -1:
+                                print('Could not find answer: %s vs. %s' %
+                                      (actual_text, cleaned_answer_text))
+                                continue
+                        else:
+                            start_position = -1
+                            end_position = -1
+                            orig_answer_text = ''
+
+                    example = SquadExample(
+                        qas_id=qas_id,
+                        question_text=question_text,
+                        doc_tokens=doc_tokens,
+                        example_id=example_id,
+                        orig_answer_text=orig_answer_text,
+                        start_position=start_position,
+                        end_position=end_position,
+                        is_impossible=is_impossible)
+                    examples.append(example)
+
+                    example_id += 1
+
+        return examples
+
+    def whitespace_tokenize(self, text):
+        """Runs basic whitespace cleaning and splitting on a piece of text."""
+        text = text.strip()
+        if not text:
+            return []
+        tokens = text.split()
+        return tokens
+
+
+class SQuADTransform(object):
+    """Dataset Transformation for BERT-style QA.
+
```

Comment:
Need to document down the transformations in the API doc. Let's try to make these code reusable and eventually merge back as gluonnlp APIs

Reply:
Please help review the documentation comments.